DEF-3265 Script Instance Context

danengelbrecht-king opened the pull request:
Adds meta functions for getting and setting instance local context values in GameObject, GUI and Render using a common API.

Updated the script callback helpers to allow automatic freeing of refs/data when a script instance is deleted.

danengelbrecht-king committed:
DEF-1183 Add meta functions and helpers for instance context values

danengelbrecht-king committed:
DEF-1183 Add context table for GO, GUI and Render script instances

danengelbrecht-king committed:
Tests Script Instance Context for GO, GUI, Render

Merged m_RefTableReference in GUI instance with m_ContextTableReference, now used for both
Instance context data and instance local references.

danengelbrecht-king committed:
Removed dmScript::SetValueToTable and dmScriptGetValueFromTable

danengelbrecht-king committed:
Simpler api for getting instance context values

Now only one meta function to implement - you return
a lua table on the stack and the helpers in dmScript does
the actual set/get of values in the table.

This opens up for us to add utility functions in dmScript
for adding instance local Refs without requiring more work
in the different systems (GO/GUI/Render)

mathiaswking reviewed (commented):
Looks good!
Only some minor suggestions

danengelbrecht-king committed:
Permit negative stack size change with DM_LUA_STACK_CHECK

danengelbrecht-king committed:
Review cleanups

danengelbrecht-king committed:
Callback data can now be instance local

Instances meta function for context table now returns ref in global lua index table
Refactored dmScript::(Un)RegisterCallback to dmScript::Create/DeleteCallback which only uses instance local Lua memory that is automatically reclaimed on script instance deletion
Moved legacy version of dmScript callback API to facebook_gameroom.cpp due to internal knowledge of data types

danengelbrecht-king committed:
Typo in iap_gameroom IAP constructor

danengelbrecht-king committed:
Use balanced luaL_ref and luaL_unref in Lua callback helpers

danengelbrecht-king committed:
Make sure to properly restore current instance when invoking callback

danengelbrecht-king committed:
Fixed up iap_gameroom.cpp

danengelbrecht-king committed:
Docs improvement, cleanups and better Lua ref counting

danengelbrecht-king committed:
Reverted unneeded change

danengelbrecht-king committed:
Fixes for facebook_gameroom.cpp

danengelbrecht-king committed:
Improved docs around callback utility functions

danengelbrecht-king committed:
Review cleanups - less DM_LUA_ERROR and more dmLogWarning
